Add new module cs_storage_pool

netservers opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Add a new CloudStack module for adding/updating/removing Primary Storage Pools.

This is part of a set of new modules that will also allow management of:
- Physical Networks
- TrafficTypes
- Storage IP Ranges
- Primary Storage Pools
- VLAN IP Ranges
- Console SSL Certificates
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
lib/ansible/modules/cloud/cloudstack/cs_storage_pool.py
##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.4.0
  config file = /home/john/.ansible.cfg
  configured module search path = [u'/home/john/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible-2.4.0-py2.7.egg/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.13 (default, May 10 2017, 20:04:28) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]

```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->
Sample playbook:
<!--- Paste verbatim command output below, e.g. before and after your change -->
```
---
- name: Setup Cloudstack
  hosts: host1
  connection: local
  tasks:
    - name: Configure RBD pool
      cs_storage_pool:
        name: Main RBD
        zone: Zone01
        pod: Pod01
        cluster: Cluster01
        scope: CLUSTER
        hypervisor: KVM
        storage_url: rbd://admin:SECRET@ceph-dns-name.example.com/rbd
        tags: rbd
```

Commit:
Add new module cs_storage_pool

resmo commented on lib/ansible/modules/cloud/cloudstack/cs_storage_pool.py:
```diff
@@ -0,0 +1,399 @@
+    returned: success
+    type: string
+    sample: a3fca65a-7db1-4891-b97c-48806a978a96
+zone_name:
```
rename to `zone`

netservers replied:
OK. Will update shortly.

resmo commented on lib/ansible/modules/cloud/cloudstack/cs_storage_pool.py:
```diff
@@ -0,0 +1,399 @@
+    returned: when scope is cluster
+    type: string
+    sample: a3fca65a-7db1-4891-b97c-48806a978a96
+cluster_name:
```
rename to `cluster`

netservers replied:
OK. Will update shortly.

resmo commented on lib/ansible/modules/cloud/cloudstack/cs_storage_pool.py:
```diff
@@ -0,0 +1,399 @@
+    returned: when scope is cluster
+    type: string
+    sample: a3fca65a-7db1-4891-b97c-48806a978a96
+pod_name:
```
rename to pod

netservers replied:
OK. Will update shortly.

resmo commented on lib/ansible/modules/cloud/cloudstack/cs_storage_pool.py:
```diff
@@ -0,0 +1,399 @@
+tags:
+    description: the Tags for the storage pool
+    returned: success
+    type: string
```
imho this is a list isn't it?

netservers replied:
Yes.

resmo commented on lib/ansible/modules/cloud/cloudstack/cs_storage_pool.py:
```diff
@@ -0,0 +1,399 @@
+        self.returns = {
+            'id': 'id',
+            'capacityiops': 'capacity_iops',
+            'zonename': 'zone_name',
```
we already return zone in super class, please remove, see https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/cloudstack.py#L434 (you need probably a rebase to devel to see it)

netservers replied:
OK

resmo commented on lib/ansible/modules/cloud/cloudstack/cs_storage_pool.py:
```diff
@@ -0,0 +1,399 @@
+            'capacityiops': 'capacity_iops',
+            'zonename': 'zone_name',
+            'zoneid': 'zone_id',
+            'podname': 'pod_name',
```
call this just pod, identical to the param we pass. Fixes inconsistency with other modules.

netservers replied:
OK

resmo commented on lib/ansible/modules/cloud/cloudstack/cs_storage_pool.py:
```diff
@@ -0,0 +1,399 @@
+            'zoneid': 'zone_id',
+            'podname': 'pod_name',
+            'podid': 'pod_id',
+            'clustername': 'cluster_name',
```
call this just cluster, identical to the param we pass. Fixes inconsistency with other modules.

netservers replied:
OK

resmo commented on lib/ansible/modules/cloud/cloudstack/cs_storage_pool.py:
```diff
@@ -0,0 +1,399 @@
+        cluster=dict(default=None),
+        scope=dict(default=None, choices=['ZONE', 'CLUSTER']),
+        hypervisor=dict(default=None),
+        provider=dict(default='DefaultPrimary'),
```
hmm is there a list of choices we can limit this param to?

netservers replied:
I don't think we can. The API docs don't give a list, and I suspect that this might be a pluggable system whereby any third party could write their own provider. Having said that I've done no research into this.

resmo replied:
alright. let's keep it as is.

resmo commented on lib/ansible/modules/cloud/cloudstack/cs_storage_pool.py:
```diff
@@ -0,0 +1,399 @@
+            'id': 'id',
+            'capacityiops': 'capacity_iops',
+            'zonename': 'zone_name',
+            'zoneid': 'zone_id',
```
I usually do not return ids of elements, except main ID. I don't see any benefit in this id, do you?

netservers replied:
OK. I know we don't really want to know it in ansible, but it might be useful for external scripting. I'll get rid of it.

resmo commented on lib/ansible/modules/cloud/cloudstack/cs_storage_pool.py:
```diff
@@ -0,0 +1,399 @@
+            'zonename': 'zone_name',
+            'zoneid': 'zone_id',
+            'podname': 'pod_name',
+            'podid': 'pod_id',
```
I usually do not return ids of elements, except main ID. I don't see any benefit in this id, do you?

netservers replied:
OK

resmo commented on lib/ansible/modules/cloud/cloudstack/cs_storage_pool.py:
```diff
@@ -0,0 +1,399 @@
+            'podname': 'pod_name',
+            'podid': 'pod_id',
+            'clustername': 'cluster_name',
+            'clusterid': 'cluster_id',
```
I usually do not return ids of elements, except main ID. I don't see any benefit in this id, do you?

netservers replied:
OK

resmo commented on lib/ansible/modules/cloud/cloudstack/cs_storage_pool.py:
```diff
@@ -0,0 +1,399 @@
+            'tags': 'tags',
+        }
+        self.storage_pool = None
+        self.zone = None
```
this is already inherited from super class

netservers replied:
OK

resmo commented on lib/ansible/modules/cloud/cloudstack/cs_storage_pool.py:
```diff
@@ -0,0 +1,399 @@
+
+    def _create_storage_pool(self):
+
+        cluster = self.module.params.get('cluster')
```
can be simplified to 
~~~
cluster = self.get_cluster(key='id')
~~~
and following lines can be removed
~~~
if cluster:
    args['clusterid'] = self.get_cluster(key='id')
~~~

netservers replied:
Am I right in assuming that if args['clusterid'] == None, clusterid will be excluded from the API request? If so, I'll just assign self.get_cluster(key='id') to args['clusterid'].

resmo replied:
yes, None values are excluded by the lib cs underneath

resmo commented on lib/ansible/modules/cloud/cloudstack/cs_storage_pool.py:
```diff
@@ -0,0 +1,399 @@
+
+        cluster = self.module.params.get('cluster')
+        scope = self.module.params.get('scope')
+        pod = self.module.params.get('pod')
```
identical here for pod,

netservers replied:
OK

resmo commented on lib/ansible/modules/cloud/cloudstack/cs_storage_pool.py:
```diff
@@ -0,0 +1,399 @@
+
+        self.result['changed'] = True
+
+        id = None
```
seems not used later in code

netservers replied:
Agreed. It's gone.

resmo reviewed (changes requested):
Thanks for your work. 

IMHO the module does not handle creating a disabled storage pool, does it?

Commit:
- Changed zone_name to name
- Changed cluster_name to cluster
- Changed pod_name to pod
- Corrected tags type in docs
- Remove unneeded returns
- Other simplifications

netservers commented:
This is still showing as changes requested, but AFAIK all requested changes have been applied.

resmo commented:
I am preparing for inclusion.

resmo commented:
While working on integration tests, I found a few issues, too many to comment on this PR. I decided to create additional patch on top of this PR also including the integration tests.
